### PR TITLE
[BugFix] BE crash in _try_get_next_compaction_task

### LINKED ISSUE #10204: BE crash in try_get_next_compaction_task
### Real behavior (Required)
```
*** Check failure stack trace: ***
    @          0xfcd647d  google::LogMessage::Fail()
    @          0xfcd8759  google::LogMessage::SendToLog()
    @          0xfcd5ff9  google::LogMessage::Flush()
    @          0xfcd8d59  google::LogMessageFatal::~LogMessageFatal()
    @          0xe0c93c7  starrocks::CompactionScheduler::_try_get_next_compaction_task()
    @          0xe0c5ac2  starrocks::CompactionScheduler::schedule()
    @          0xcd2d398  _ZZN9starrocks13StorageEngine16start_bg_threadsEvENKUlvE5_clEv
    @          0xcd3c796  _ZSt13__invoke_implIvZN9starrocks13StorageEngine16start_bg_threadsEvEUlvE5_JEET_St14__invoke_otherOT0_DpOT1_
    @          0xcd3c35c  _ZSt8__invokeIZN9starrocks13StorageEngine16start_bg_threadsEvEUlvE5_JEENSt15__invoke_resultIT_JDpT0_EE4typeEOS4_DpOS5_
    @          0xcd3bf6e  _ZNSt6thread8_InvokerISt5tupleIJZN9starrocks13StorageEngine16start_bg_threadsEvEUlvE5_EEE9_M_invokeIJLm0EEEEvSt12_Index_tupleIJXspT_EEE
    @          0xcd3bd46  _ZNSt6thread8_InvokerISt5tupleIJZN9starrocks13StorageEngine16start_bg_threadsEvEUlvE5_EEEclEv
    @          0xcd3bbae  _ZNSt6thread11_State_implINS_8_InvokerISt5tupleIJZN9starrocks13StorageEngine16start_bg_threadsEvEUlvE5_EEEEE6_M_runEv
    @         0x1196d2a0  execute_native_thread_routine
    @     0x7fd228c75ea5  start_thread
    @     0x7fd22773bb0d  __clone
    @              (nil)  (unknown)
```

### StarRocks version (Required)
2.4.0-0815-f5c2bd7

### PR DESCRIPTION
**Pull request by rickif**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #10204

## Problem Summary(Required) ：

The result `need_reschedule` and `can_do` of `CompactionScheduler::_can_do_compaction` may be both false in the origin implementation, which leads the crash in ASAN mode. This PR fix the problem.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

File: be/src/storage/compaction_scheduler.cpp [hunk 1]
```diff
@@ -206,45 +199,42 @@ std::shared_ptr<CompactionTask> CompactionScheduler::_try_get_next_compaction_ta
...
     std::vector<CompactionCandidate> tmp_candidates;
     CompactionCandidate compaction_candidate;
-    bool found = false;
     std::shared_ptr<CompactionTask> compaction_task;
+
     while (true) {
         if (!_can_schedule_next()) {
             VLOG(2) << "_can_schedule_next is false. skip";
-            break;
+            return nullptr;
```
Comment (meegoo) [inline be/src/storage/compaction_scheduler.cpp:207]: should not return directly since `tmp_candidates` haven't insert to compaction_manager

**Review by meegoo — CHANGES_REQUESTED** [review]

File: be/src/storage/compaction_scheduler.cpp (hunk 1 above)
Comment (rickif) [reply be/src/storage/compaction_scheduler.cpp:207]: `tmp_candidates` is empty here, so it's ok not to insert the `tmp_candidates` to compaction manager.

Comment (meegoo) [reply be/src/storage/compaction_scheduler.cpp:207]: it's a while loop, tmp_candidates may be inserted before this round?

Comment (rickif) [reply be/src/storage/compaction_scheduler.cpp]: Updated.

File: be/src/storage/compaction_scheduler.cpp [hunk 2]
```diff
@@ -224,27 +217,24 @@ std::shared_ptr<CompactionTask> CompactionScheduler::_try_get_next_compaction_ta
...
         VLOG(2) << "try tablet:" << tablet_id << ", compaction type:" << to_string(compaction_type);
-        bool need_reschedule = true;
-        found = _can_do_compaction(compaction_candidate, &need_reschedule, &compaction_task);
-        if (need_reschedule) {
-            tmp_candidates.emplace_back(std::move(compaction_candidate));
-        } else {
-            DCHECK(found);
+
+        if (!_check_precondition(compaction_candidate)) {
+            LOG(INFO) << "check compaction precondition failed, candidate info: " << compaction_candidate.to_string();
```
Comment (ABingHuang) [inline be/src/storage/compaction_scheduler.cpp:222]: there may be many log here. consider to use VLOG?

**Review by ABingHuang — DISMISSED** [review]
LGTM

File: be/src/storage/compaction_scheduler.cpp (hunk 2 above)
Comment (rickif) [reply be/src/storage/compaction_scheduler.cpp]: Updated.

**Review by meegoo — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Review by chaoyli — APPROVED** [review]

**Comment by rickif** [comment]
@mergifyio backport branch-2.4

**Pull request merged**